src/upf/src/upf.cpp

```diff
@@ -917,99 +1241,80 @@ bool eval_upf(sta::dbNetwork* network, utl::Logger* logger, odb::dbBlock* block)
-  // cell into either first or second domain?? (Not sure exactly)
+  // get all cell names for level shifter
+  // TODO: should be replaced later by querying lib to find if inst is level
+  // shifter
+  std::unordered_map<std::string, bool> level_shifter_cells;
+  auto shifters = block->getLevelShifters();
+  for (auto&& shifter : shifters) {
+    level_shifter_cells[shifter->getCellName()] = true;
+  }
 
   for (auto&& domain : pds) {
     if (domain == top_domain) {
       continue;
     }
 
-    // For now we're only using the first isolation
-    // TODO: determine what needs to be done in case of multiple strategies
-    // for same domain
-    auto isos = domain->getIsolations();
-
-    if (isos.empty()) {
-      continue;
-    }
-
-    if (isos.size() > 1) {
-      logger->warn(
-          utl::UPF,
-          30,
-          "Multiple isolation strategies defined for the same power domain %s.",
-          domain->getName());
-    }
-
-    odb::dbIsolation* iso = isos[0];
-
-    odb::dbMTerm* enable_term = nullptr;
-    odb::dbMTerm* data_term = nullptr;
-    odb::dbMTerm* output_term = nullptr;
-    odb::dbMaster* smallest_iso_m = nullptr;
-    bool invert_output, invert_control;
-
-    if (!find_smallest_isolation(network,
-                                 logger,
-                                 iso,
-                                 smallest_iso_m,
-                                 enable_term,
-                                 data_term,
-                                 output_term,
-                                 invert_output,
-                                 invert_control)) {
-      continue;
-    }
-
-    if ((invert_output || invert_control) && !inverter_found) {
-      logger->warn(utl::UPF, 31, "can't find any inverters");
-      continue;
-    }
-
-    // Iterate through all pd's instances and determine if any signal needs
-    // isolation If one is found, create dbInst from the isolation cell +
-    // connect control signal of isolation to dbInst
-    // connect previous net to data input
-    // connect output from dbInst to
-
     odb::dbGroup* curr_group = domain->getGroup();
     auto insts = curr_group->getInsts();
 
     for (auto&& inst : insts) {
+      if (level_shifter_cells[inst->getMaster()->getName()]) {
+        continue;
+      }
       auto iterms = inst->getITerms();
       for (auto&& iterm : iterms) {
-        if (iterm->getIoType() != odb::dbIoType::OUTPUT) {
-          continue;
+        // get all connected iterms as well
+        auto connected_iterms = get_connected_terms(block, iterm);
+        for (auto&& connected_iterm : connected_iterms) {
+          odb::dbITerm* target_iterm = connected_iterm.first;
+          odb::dbPowerDomain* target_domain = connected_iterm.second;
+
+          // if target instance is a level shifter then skip
+          if (level_shifter_cells
+                  [target_iterm->getInst()->getMaster()->getName()]) {
+            continue;
+          }
```

Comment:
will this skip adding an isolation cell before a level shifter?

Reply:
you can't have both level shifters and isolation cells according to the code I've written. If the voltage between both domains are equal and the instance is a driver then and only then isolate, otherwise check if any shifting strategies apply.